### PR TITLE
copr: add non-vectorized uuid

### PR DESCRIPTION
**Pull request by chacha923**
Signed-off-by: chacha923 chacha923@aliyun.com

PCP #5751

###  What have you changed?

Add non-vectorized uuid

###  What is the type of the changes?

- New feature (a change which adds functionality)

###  How is the PR tested?

- Unit test

###  Does this PR affect documentation (docs) or should it be mentioned in the release notes?
no

###  Does this PR affect `tidb-ansible`?
no

###  Refer to a related PR or issue link (optional)
#3275

### CODE REVIEW

**Comment by chacha923** [comment]
Signed-off-by: chacha923 chacha923@aliyun.com

File: components/tidb_query/Cargo.toml [hunk 1]
```diff
@@ -48,6 +48,7 @@ tipb = { git = "https://github.com/pingcap/tipb.git", default-features = false }
 
 tipb_helper = { path = "../tipb_helper" }
 twoway = "0.2.0"
+uuid = { version = "0.7", features = [ "serde", "v4" ] }
```
Comment (lonng) [inline components/tidb_query/Cargo.toml]: Why not use the latest version (0.8.1)?

File: components/tidb_query/src/expr/builtin_miscellaneous.rs [hunk 2]
```diff
@@ -186,6 +187,17 @@ impl ScalarFunc {
...
         }
     }
+
+    pub fn uuid<'a, 'b: 'a>(
+        &'b self,
+        _ctx: &mut EvalContext,
+        _row: &[Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let result = Uuid::new_v4();
+        let mut buf = [b'!'; 40];
```
Comment (lonng) [inline components/tidb_query/src/expr/builtin_miscellaneous.rs]: Seems 32 bytes is enough.

File: components/tidb_query/src/expr/builtin_miscellaneous.rs [hunk 3]
```diff
@@ -186,6 +187,17 @@ impl ScalarFunc {
...
+        let uuid_str = result.to_hyphenated().encode_lower(&mut buf);
+        Ok(Some(Cow::Owned(uuid_str.as_bytes().to_vec())))
```
Comment (lonng) [inline components/tidb_query/src/expr/builtin_miscellaneous.rs]: We can use the `buf` directly.

File: components/tidb_query/Cargo.toml [hunk 4]
```diff
@@ -49,6 +49,7 @@ tipb = { git = "https://github.com/pingcap/tipb.git", default-features = false }
 
 tipb_helper = { path = "../tipb_helper" }
 twoway = "0.2.0"
+uuid = { version = "0.8.1", features = [ "serde", "v4" ] }
```
Comment (sticnarf) [inline components/tidb_query/Cargo.toml]: Do we need the `serde` feature?

Comment (chacha923) [reply components/tidb_query/Cargo.toml]: > Do we need the `serde` feature?

really not need serde, i delete it

File: components/tidb_query/src/expr/builtin_miscellaneous.rs [hunk 5]
```diff
@@ -186,6 +187,17 @@ impl ScalarFunc {
...
         }
     }
+
+    pub fn uuid<'a, 'b: 'a>(
+        &'b self,
+        _ctx: &mut EvalContext,
+        _row: &[Datum],
+    ) -> Result<Option<Cow<'a, [u8]>>> {
+        let result = Uuid::new_v4();
+        let mut buf = [b'!'; uuid::adapter::Hyphenated::LENGTH];
```
Comment (sticnarf) [inline components/tidb_query/src/expr/builtin_miscellaneous.rs]: Just another tiny problem.
I think we can simply initial buf using `vec![0; uuid::adapter::Hyphenated::LENGTH]`. Then we don't need to fill the buf at stack and then copy them to the heap.

Comment (chacha923) [reply components/tidb_query/src/expr/builtin_miscellaneous.rs]: ok, i use vec!

**Pull request merged**